src/kani-verifier/src/lib.rs

```diff
@@ -0,0 +1,239 @@
+
+#![warn(clippy::all, clippy::cargo)]
+
+use std::env;
+use std::ffi::OsString;
+use std::os::unix::prelude::CommandExt;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+use anyhow::{bail, Result};
+
+/// Comes from our Cargo.toml manifest file. Must correspond to our release verion.
+const VERSION: &str = env!("CARGO_PKG_VERSION");
+/// Set by our `build.rs`, reflects the Rust target triple we're building for
+const TARGET: &str = env!("TARGET");
+
+/// Typically `~/.kani/kani-1.x/`
+fn kani_dir() -> PathBuf {
+    home::home_dir()
+        .expect("Couldn't find home dir?")
+        .join(".kani")
+        .join(format!("kani-{}", VERSION))
+}
+
+/// The filename of the release bundle
+fn download_filename() -> String {
+    format!("kani-{}-{}.tar.gz", VERSION, TARGET)
+}
+
+/// Helper to find the download URL for this version of Kani
+fn download_url() -> String {
+    let tag: &str = &format!("kani-{}", VERSION);
+    let file: &str = &download_filename();
+    format!("https://github.com/model-checking/kani/releases/download/{}/{}", tag, file)
+}
+
+/// Effectively the entry point (i.e. `main` function) for both our proxy binaries.
+pub fn proxy(bin: &str) -> Result<()> {
+    // In an effort to keep our dependencies minimal, we do the bare minimum argument parsing
+    let args: Vec<_> = std::env::args_os().collect();
+    if args.len() >= 2 && args[1] == "setup" {
+        if args.len() >= 4 && args[2] == "--use-local-bundle" {
+            setup(Some(args[3].clone()))
+        } else {
+            setup(None)
+        }
+    } else {
+        if !appears_setup() {
+            setup(None)?;
+        }
+        exec(bin)
+    }
+}
+
+/// Fast check to see if we look setup already
+fn appears_setup() -> bool {
+    kani_dir().exists()
+}
+
+/// Sets up Kani by unpacking/installing to `~/.kani/kani-VERSION`
+fn setup(use_local_bundle: Option<OsString>) -> Result<()> {
+    // e.g. `~/.kani/`
+    let kani_dir = kani_dir();
+    let base_dir = kani_dir.parent().expect("No base directory?");
+
+    println!("Running Kani first-time setup...");
+
+    // 0. Make sure we have a `~/.kani/`
+    std::fs::create_dir_all(&base_dir)?;
+
+    // 1. Download and unpack our release binary
+    if let Some(pathstr) = use_local_bundle {
+        let path = Path::new(&pathstr).canonicalize()?;
+        Command::new("tar").arg("zxf").arg(&path).current_dir(base_dir).run()?;
+
+        // when given a bundle, it's often "-latest" but we expect "-1.0" or something. Hack it up.
+        let file = path.file_name().expect("has filename").to_string_lossy();
+        let components: Vec<_> = file.split("-").collect();
+        let expected_dir = format!("{}-{}", components[0], components[1]);
+
+        std::fs::rename(base_dir.join(expected_dir), &kani_dir)?;
+    } else {
+        let bundle = base_dir.join(download_filename());
+        Command::new("curl").args(&["-sSLf", "-o"]).arg(&bundle).arg(download_url()).run()?;
```

Comment:
Can we add some error handling here? We probably want to give the user a nice message if we fail to download the bundle. And a message that help us root cause the issue. :)

Reply:
Sure, I will do a whole pass trying to improve the error handling and messaging. I know it's a problem currently.

It's not suppressing the output though, so you do see a `curl` error here saying 404 and giving the URL, fwiw.